core/src/main/scala/org/apache/spark/deploy/SparkS3Util.scala

```diff
@@ -0,0 +1,299 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy
+
+import java.io.IOException
+import java.net.URI
+import java.util
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import com.amazonaws.{AmazonClientException, ClientConfiguration, Protocol}
+import com.amazonaws.auth.{AWSCredentialsProvider, BasicAWSCredentials, InstanceProfileCredentialsProvider, STSAssumeRoleSessionCredentialsProvider}
+import com.amazonaws.internal.StaticCredentialsProvider
+import com.amazonaws.services.s3.AmazonS3Client
+import com.amazonaws.services.s3.model.{ListObjectsRequest, ObjectListing, S3ObjectSummary}
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{BlockLocation, FileStatus, LocatedFileStatus, Path, PathFilter, RemoteIterator}
+import org.apache.hadoop.fs.s3.S3Credentials
+import org.apache.hadoop.io.compress.{CompressionCodecFactory, SplittableCompressionCodec}
+import org.apache.hadoop.mapred.{FileSplit, InputSplit, JobConf}
+
+import org.apache.spark.{SparkConf, Logging}
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.util.Utils
+
+import com.google.common.base.{Preconditions, Strings}
+import com.google.common.cache.{Cache, CacheBuilder}
+import com.google.common.collect.AbstractSequentialIterator
+
+/**
+ * :: DeveloperApi ::
+ * Contains util methods to interact with S3 from Spark.
+ */
+@DeveloperApi
+object SparkS3Util extends Logging {
+  private val sparkConf = new SparkConf()
+  val conf: Configuration = SparkHadoopUtil.get.newConfiguration(sparkConf)
+
+  private val s3ClientCache: Cache[String, AmazonS3Client] = CacheBuilder
+    .newBuilder
+    .concurrencyLevel(Runtime.getRuntime.availableProcessors)
+    .build[String, AmazonS3Client]
+
+  private val S3_CONNECT_TIMEOUT: String = "spark.s3.connect.timeout"
+  private val S3_MAX_CONNECTIONS: String = "spark.s3.max.connections"
+  private val S3_MAX_ERROR_RETRIES: String = "spark.s3.max.error.retries"
+  private val S3_SOCKET_TIMEOUT: String = "spark.s3.socket.timeout"
+  private val S3_SSL_ENABLED: String = "spark.s3.ssl.enabled"
+  private val S3_USE_INSTANCE_CREDENTIALS: String = "spark.s3.use.instance.credentials"
+
+  private val FILEINPUTFORMAT_INPUTDIR: String =
+    "mapreduce.input.fileinputformat.inputdir"
+  private val FILEINPUTFORMAT_NUMINPUTFILES: String =
+    "mapreduce.input.fileinputformat.numinputfiles"
+  private val FILEINPUTFORMAT_SPLIT_MINSIZE: String =
+    "mapreduce.input.fileinputformat.split.minsize"
```

Comment:
Do we need to create those `val`s at here? Are they available as public static fields in `FileInputFormat`? Also, to make it work with both hadoop 1 and hadoop 2, do we need to set both versions of those flags?